lambda/aws-qa-appsync-opensearch/question_answering/src/qa_agent/image_qa.py

```diff
@@ -0,0 +1,258 @@
+metrics = Metrics(namespace="question_answering", service="QUESTION_ANSWERING")
+
+bucket_name = os.environ['INPUT_BUCKET']
+
+def run_qa_agent_on_image_no_memory(input_params):
+    logger.info("starting qa agent without memory on uploaded image")
+
+    status_variables = {
+        'jobstatus': JobStatus.WORKING.status,
+        'answer': JobStatus.WORKING.get_message(),
+        'jobid': input_params['jobid'],
+        'filename': input_params['filename'],
+        'question': input_params['question'],
+        'sources': ['']
+    }
+
+    qa_model=input_params['qa_model']
+    
+    if 'modelId' in qa_model:
+         qa_modelId= qa_model['modelId']
+    else:
+         logger.error(' No model id defined, cannot answer question , returning...')
+         status_variables['jobstatus'] = JobStatus.ERROR_LOAD_LLM.status
+         status_variables['answer'] = JobStatus.ERROR_LOAD_LLM.status
+         send_job_status(status_variables)
+         return status_variables
+
+
+    base64_bytes = input_params['question'].encode("utf-8")
+    sample_string_bytes = base64.b64decode(base64_bytes)
+    decoded_question = sample_string_bytes.decode("utf-8")
+    logger.info(decoded_question)
+
+    filename = input_params['filename']
+    image_url = input_params['presignedurl']
+    model_provider = input_params['qa_model']['provider']
+    logger.info(f"Generating response to question for file {filename} image_url {image_url}")
+    
+    status_variables['sources'] = [filename]
+    if image_url is None:
+        status_variables['jobstatus'] = JobStatus.ERROR_LOAD_DOC.status
+        error = JobStatus.ERROR_LOAD_DOC.get_message()
+        status_variables['answer'] = error.decode("utf-8")
+        send_job_status(status_variables)
+        return status_variables
+
+    streaming = input_params.get("streaming", False)
+    callback_manager = [StreamingCallbackHandler(status_variables)] if streaming else None  
+    if model_provider=='Sagemaker Endpoint':
+        _qa_llm = MultiModal.sagemakerendpoint_llm(qa_modelId)
+    elif model_provider=='Bedrock':
+        _qa_llm = get_llm(callback_manager)
+    else:
+        logger.error('Invalid Model provider, cannot load  LLM , returning..')
+        status_variables['jobstatus'] = JobStatus.ERROR_LOAD_LLM.status
+        error = JobStatus.ERROR_LOAD_LLM.get_message()
+        status_variables['answer'] = error.decode("utf-8")
+        send_job_status(status_variables)
+        return status_variables
+    
+    template = """\n\nUser: {question}![]({image})<end_of_utterance>
+         \n\nAssistant:"""
+    prompt = PromptTemplate(template=template, input_variables=["image", "question"])
+    chain = LLMChain(llm=_qa_llm, prompt=prompt, verbose=input_params['verbose'])
+
+    try:
+       
+        logger.info(f'decoded_question is: {decoded_question}')
+        tmp = chain.predict(image=image_url, question=decoded_question)
+        answer=tmp.split("Assistant:",1)[1]
```

Comment:
is this the output parsed for a specific model ?

Reply:
yes this is only for Idefics on sagemaker endpoint. Once tested with more model will make it more generic.